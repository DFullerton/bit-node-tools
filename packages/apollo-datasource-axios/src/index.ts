export {
  ApolloDataSourceConfig,
  ApolloAxiosInstance,
  ApolloAxiosPromise,
  ApolloAxiosResponse,
  AxiosDataSource,
} from './AxiosDataSource';
export { cacheAdapter } from './cacheAdapter';
