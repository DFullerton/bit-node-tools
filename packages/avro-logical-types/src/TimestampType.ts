import { Type, types } from 'avsc';
/**
 * Custom logical type used to encode native Date objects as long.
 */
export class TimestampType extends types.LogicalType {
  _fromValue(val: number) {
    return new Date(val).toISOString();
  }
  _toValue(date: any) {
    return date instanceof Date ? date.getTime() : date;
  }
  _resolve(type: any) {
    if (Type.isType(type, 'long', 'logical:timestamp-millis')) {
      return this._fromValue;
    }
    return undefined;
  }
}
