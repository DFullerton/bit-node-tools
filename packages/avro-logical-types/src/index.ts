export { DateType } from './DateType';
export { TimestampType } from './TimestampType';
