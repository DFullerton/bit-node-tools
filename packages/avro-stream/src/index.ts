export { AvroDeserializer } from './AvroDeserializer';
export { AvroSerializer } from './AvroSerializer';
export { SchemaRegistryResolver } from './SchemaRegistryResolver';
export { SchemaResolver, AvroMessage, AvroProduceRequest } from './types';
export { constructMessage, deconstructMessage } from './message';
