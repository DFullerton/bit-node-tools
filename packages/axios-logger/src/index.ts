import { AxiosError, AxiosRequestConfig, AxiosResponse } from 'axios';

export interface AxiosLogger {
  redact?: string[];
  requestStartedAt?: Date;
}

export interface LogMeta {
  uri?: string;
  method?: string;
  requestBody?: any;
  responseBody?: any;
  params: any;
  status: number;
  responseTime: number;
}

export const startHeader = 'X-Request-Started';
export const redactHeader = 'X-Redact-Log';

const redactPath = (path: string[], obj: any): any => {
  const [current, ...rest] = path;

  return current === '*' && Array.isArray(obj)
    ? obj.map(item => redactPath(rest, item))
    : obj[current]
    ? { ...obj, [current]: rest.length ? redactPath(rest, obj[current]) : '<Removed>' }
    : obj;
};

const defaultRedact = ['requestBody', 'uri', 'params', 'responseBody'];

export const redactPaths = (paths: string[], object: any) =>
  paths.reduce((current, path) => redactPath(path.split('.'), current), object);

export const getMeta = (response: AxiosResponse): LogMeta => {
  const { url: uri, data, params, method, headers } = response.config;
  const { data: responseBody, status } = response;
  const requestBody = data === undefined ? undefined : JSON.parse(data);
  const responseTime = headers[startHeader] ? new Date().getTime() - headers[startHeader].getTime() : undefined;
  const meta = { uri, method, requestBody, responseBody, params, status, responseTime };

  return redactPaths(
    headers[redactHeader] ? headers[redactHeader].split(',').map((path: string) => path.trim()) : defaultRedact,
    meta,
  );
};

export const axiosLogger = (log: (level: 'info' | 'error', meta: LogMeta, config: AxiosRequestConfig) => void) => ({
  request: {
    onFulfilled: (config: AxiosRequestConfig): AxiosRequestConfig => ({
      ...config,
      headers: {
        ...config.headers,
        [startHeader]: config.headers[startHeader] || new Date(),
      },
    }),
  },
  response: {
    onFulfilled: (response: AxiosResponse) => {
      log('info', getMeta(response), response.config);
      return response;
    },
    onRejected: (error: AxiosError) => {
      if (error.response) {
        const { responseBody, ...meta } = getMeta(error.response);
        log('error', meta, error.config);
      }
      return Promise.reject(error);
    },
  },
});
