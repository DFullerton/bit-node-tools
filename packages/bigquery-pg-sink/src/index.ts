export { BigQueryPGSinkStream } from './PGSinkStream';
export { PGSinkStreamOptions, InsertBatch } from './types';
