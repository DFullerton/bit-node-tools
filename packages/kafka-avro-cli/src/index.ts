export {
  NullWritable,
  ConsumerProgressTransform,
  LogConsumerProgressTransform,
  LogProducerTransform,
  FileWritable,
  FileReadable,
  StopOnFinishProgressTransform,
} from './stream';
export { Args, Config, ConsumeArgs, CreateTopicArgs, ProduceArgs, SchemaArgs, TopicArgs, filterSearch } from './cli';
