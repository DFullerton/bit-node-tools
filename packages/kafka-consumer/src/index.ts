import { DateType, TimestampType } from '@ovotech/avro-logical-types';
import { AvroDeserializer, AvroMessage } from '@ovotech/avro-stream';
import { rePipeline } from '@ovotech/re-pipeline';
import { Logger } from '@ovotech/winston-logger';
import { ConsumerGroupStream, ConsumerGroupStreamOptions } from 'kafka-node';
import { Writable } from 'stream';

interface KafkaConsumerInput {
  OPTIONS: ConsumerGroupStreamOptions;
  KAFKA_SCHEMA_REGISTRY: string;
  KAFKA_TOPICS: string | string[];
}

/**
 * Create and maintain a kafka notification consumer
 *
 * @param env
 * @param logger
 * @param onDetectNotification function to process a given notification
 */
export const createKafkaConsumer = (
  { OPTIONS, KAFKA_SCHEMA_REGISTRY, KAFKA_TOPICS }: KafkaConsumerInput,
  logger: Logger,
  onDetectNotification: (message: AvroMessage) => Promise<void>,
) => {
  const avroDeserializer = new AvroDeserializer(KAFKA_SCHEMA_REGISTRY!, {
    logicalTypes: { date: DateType, 'timestamp-millis': TimestampType },
  });

  const consumerStream = new ConsumerGroupStream(OPTIONS, KAFKA_TOPICS);

  const notificationConsumer = new Writable({
    objectMode: true,
    write: async (message: AvroMessage, _, callback) => {
      // Await the processing of the notification
      await onDetectNotification(message);
      callback();
    },
  });

  rePipeline(error => logger.error(error.message, { error }), consumerStream, avroDeserializer, notificationConsumer);

  return consumerStream;
};
