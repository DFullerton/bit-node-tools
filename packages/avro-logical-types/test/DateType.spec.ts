import { Schema, Type } from 'avsc';
import { DateType } from '../src';

describe('Unit test', () => {
  it('DateType', async () => {
    const type = Type.forSchema(
      {
        type: 'record',
        fields: [{ name: 'startedAt', type: { type: 'int', logicalType: 'date' } }],
      } as Schema,
      { logicalTypes: { date: DateType } },
    );

    const values = [{ startedAt: new Date('2018-06-05') }, { startedAt: 17687 }];
    const expected = [{ startedAt: '2018-06-05T00:00:00.000Z' }, { startedAt: '2018-06-05T00:00:00.000Z' }];

    const buffers = values.map(value => type.toBuffer(value));
    const converted = buffers.map(buffer => type.fromBuffer(buffer));

    expect(buffers[0]).toEqual(buffers[1]);
    expect(converted).toEqual(expected);
    expect(buffers).toMatchSnapshot();
  });

  it('Test resolve from int', () => {
    const type = Type.forSchema(
      {
        type: 'record',
        fields: [{ name: 'startedAt', type: { type: 'int', logicalType: 'date' } }],
      } as Schema,
      { logicalTypes: { date: DateType } },
    );

    const longType = Type.forSchema({
      type: 'record',
      fields: [{ name: 'startedAt', type: 'int' }, { name: 'name', type: 'string' }],
    } as Schema);

    const expected = { startedAt: '2018-06-05T00:00:00.000Z' };
    const longBuffer = longType.toBuffer({ startedAt: 17687, name: 'test' });
    const resolver = type.createResolver(longType);
    const value = type.fromBuffer(longBuffer, resolver);

    expect(value).toEqual(expected);
  });

  it('Test invalid resolver', () => {
    const type = Type.forSchema(
      {
        type: 'record',
        fields: [{ name: 'startedAt', type: { type: 'int', logicalType: 'date' } }],
      } as Schema,
      { logicalTypes: { date: DateType } },
    );

    const longType = Type.forSchema({
      type: 'record',
      fields: [{ name: 'startedAt', type: 'string' }],
    } as Schema);

    expect(() => type.createResolver(longType)).toThrowError();
  });
});
