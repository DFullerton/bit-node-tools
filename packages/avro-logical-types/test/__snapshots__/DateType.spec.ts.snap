// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test DateType 1`] = `
Array [
  Object {
    "data": Array [
      174,
      148,
      2,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [
      174,
      148,
      2,
    ],
    "type": "Buffer",
  },
]
`;
